### PR TITLE
Option to select extensions compile-time. Fix #50 #123

### LINKED ISSUE #50: Compile two separate binaries, one with http plugin and one without
For some people having networking capabilities extend the attack/vulnerability surface.

It would be nice if we can exclude networking code in the compiled executable 
(also because keepasshttp don't have good support on Windows, See: https://groups.google.com/forum/#!topic/keepassx-reboot/yV__ujozMtI)
letting user (or disto-maintainer) decide if include keepasshttp or not in the executable

Note: by default keepasshttp is disabled at runtime

### PR DESCRIPTION
**Pull request by TheZ3ro**
## Description

This PR add Options to select extensions at compile time to let user and distributors to select with component add to KeePassXC.

The extensions/components supported right now are:
- Autotype
- KeePassHTTP
- Yubikey (still not working, waiting for #119)

You can enable (ON) or disable (OFF) extensions with the following cmake flag:
- `-DWITH_XC_HTTP=ON` for **KeePassHTTP**
- `-DWITH_XC_AUTOTYPE=ON` for **Autotype**
- `-DWITH_XC_YUBIKEY=ON` for **Yubikey**

At the end of the **cmake** command, enabled and disabled extensions will be listed:
```
-- Enabled features:
 * Autotype , Auto-type passwords in Input fields

-- Disabled features:
 * KeePassHTTP , KeePassHTTP support for ChromeIPass and PassIFox
```

Right now, by default no extension is enabled. (*We can discuss about this*)

## Motivation and Context

This fixes #50 and #123 

## How Has This Been Tested?

Manually tested the enabled and disabled option for HTTP and Autotype.
**NEED** an update to the travis file to set all the extension to ON so it will test them
(or maybe building a matrix with different ON/OFF combination?)

## Types of changes

- :negative_squared_cross_mark: Bug fix (non-breaking change which fixes an issue)
- :white_check_mark: New feature (non-breaking change which adds functionality)
- :white_check_mark: Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- :white_check_mark: I have read the **CONTRIBUTING** document. [REQUIRED]
- :white_check_mark: My code follows the code style of this project. [REQUIRED]
- :white_check_mark: All new and existing tests passed. [REQUIRED]
- :white_check_mark: My change requires a change to the documentation.
- :negative_squared_cross_mark: I have updated the documentation accordingly.
- :negative_squared_cross_mark: I have added tests to cover my changes.

### CODE REVIEW

**Comment by TheZ3ro** [comment]
Seems that cutting of the autotype cmake don't disable autotype at all.
Fixing.

**Comment by droidmonkey** [comment]
Excellent, I'll see if I can add the about dialog feature i mentioned.

**Comment by TheZ3ro** [comment]
It's already been build into a static library that it's later linked in the main executable

Implementing the AboutDialog should be easy with the `WITH_XC_*` ifdef flags

Anyway, the Autotype plugin works even if you cut off the **src/autotype/Cmake**.
I think it need a static library like the http as well

File: tests/config-keepassx-tests.h.cmake [hunk 1]
```diff
@@ -5,4 +5,8 @@
 
 #define KEEPASSX_TEST_DATA_DIR "${KEEPASSX_TEST_DATA_DIR}"
 
+#cmakedefine WITH_XC_HTTP
```
Comment (Manko10) [inline tests/config-keepassx-tests.h.cmake]: Are these defines used anywhere or is this just for when we have proper HTTP tests?

Comment (Manko10) [inline tests/config-keepassx-tests.h.cmake]: Alright, seems fair. Code's looking okay for me.

**Comment by droidmonkey** [comment]
I implemented the about dialog extensions view. It's not scalable with #ifdef's, there might be a better way to do it in the future when we define an actual plugin interface (maybe).

**Comment by TheZ3ro** [comment]
I think it's the best option we have right now

**Comment by TheZ3ro** [comment]
I think this can be merged @droidmonkey what do you think?

**Comment by droidmonkey** [comment]
I might be able to look at it this week, took some time off recently.

**Comment by droidmonkey** [comment]
I made a bunch of changes to the autotype lib building. It did not work at all on Windows, a bunch of dual definitions and the EXPORT/IMPORT was backwards (this was an error from many moons ago). I tested on Fedora and everything seemed ok, this needs to be tested on OpenSuse and Apple.

**Comment by TheZ3ro** [comment]
Tested on macOS Sierra, compile fine and works fine but testgui give some error (maybe unrelated)

```
********* Start testing of TestGui *********
Config: Using QtTest library 5.7.0, Qt 5.7.0 (x86_64-little_endian-lp64 shared (dynamic) release build; by Clang 8.0.0 (clang-800.0.38) (Apple))
QWARN  : TestGui::initTestCase() FilePath::DataPath: can't find data dir
QWARN  : TestGui::initTestCase() Unable to load auto-type plugin:
Cannot load library /Applications/KeePassXC.app/Contents/PlugIns/libkeepassx-autotype-cocoa.so: (dlopen(/Applications/KeePassXC.app/Contents/PlugIns/libkeepassx-autotype-cocoa.so, 6): Library not loaded: @executable_path/../Frameworks/QtGui.framework/Versions/5/QtGui
  Referenced from: /Applications/KeePassXC.app/Contents/Frameworks/QtWidgets.framework/Versions/5/QtWidgets
  Reason: image not found)
PASS   : TestGui::initTestCase()
QWARN  : TestGui::testMergeDatabase() QSignalSpy: Unable to handle parameter 'mergedDb' of type 'Database*' of method 'databaseMerged', use qRegisterMetaType to register it.
PASS   : TestGui::testMergeDatabase()
FAIL!  : TestGui::testAutoreloadDatabase() Compared values are not the same
   Actual   (m_db->rootGroup()->findChildByName("General")->entries().size()): 0
   Expected (1)                                                              : 1
   Loc: [/Users/thezero/keepassxc/tests/gui/TestGui.cpp(166)]
PASS   : TestGui::testTabs()
PASS   : TestGui::testEditEntry()
PASS   : TestGui::testAddEntry()
PASS   : TestGui::testEntryEntropy()
PASS   : TestGui::testSearch()
PASS   : TestGui::testDeleteEntry()
PASS   : TestGui::testCloneEntry()
PASS   : TestGui::testDragAndDropEntry()
PASS   : TestGui::testDragAndDropGroup()
PASS   : TestGui::testSaveAs()
PASS   : TestGui::testSave()
PASS   : TestGui::testDatabaseSettings()
PASS   : TestGui::testKeePass1Import()
PASS   : TestGui::testDatabaseLocking()
PASS   : TestGui::cleanupTestCase()
Totals: 17 passed, 1 failed, 0 skipped, 0 blacklisted, 11150ms
********* Finished testing of TestGui *********
```

**Comment by droidmonkey** [comment]
The autoreload tests are a little hokey since they wait an arbitrary 1.5 seconds for the auto-detection of the file change to work. This can likely be enhanced in the future, but it is definitely unrelated to this PR.

**Comment by TheZ3ro** [comment]
Really it's the autoreload that fails

**Comment by TheZ3ro** [comment]
Ping @droidmonkey

**Comment by droidmonkey** [comment]
Thanks. This pull request has been integrated manually.

**Comment by droidmonkey** [comment]
Merged using multiple squash commits to preserve authorship.

**Comment by dannysu** [comment]
I think it'd also be good to update the Wiki to reflect the additional flags.
Or how about in the message that shows up, give instruction on how to do so?

E.g.
```
-- Disabled features:
 * KeePassHTTP , KeePassHTTP support for ChromeIPass and PassIFox (Enable by -DWITH_XC_HTTP=ON)
```

**Comment by TheZ3ro** [comment]
I like the idea of showing up in the Cmake message, but it will show up also when enabled.

I will update the wiki asap (now)

**Pull request closed without merging**